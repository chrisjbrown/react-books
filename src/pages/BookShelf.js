import { useState } from "react";
import { Link, useLocation, useHistory } from "react-router-dom";
import { FaChevronDown, FaChevronUp, FaBook } from "react-icons/fa";
import { connect } from "react-redux";

import styled from "styled-components";
import Book from "../components/Book";
import Button from "../components/Button";
import Header from "../components/Header";
import AddBookModal from "../components/AddBookModal";
import { removeBook } from "../store/books/actions";

const BookList = styled.div`
  display: grid;
  grid-template-columns: ${(props) =>
    props.gridView ? "repeat(auto-fit, 160px);" : "1fr"}
  column-gap: 20px;
  row-gap: 20px;
`;

const GridControls = styled.div`
  display: flex;
  justify-content: space-between;
  padding: 10px;
  border: 1px solid black;
  margin-bottom: 20px;
`;

const StyledLink = styled(Link)`
  display: flex;
  text-decoration: none;
  &:focus,
  &:hover,
  &:visited,
  &:link,
  &:active {
    text-decoration: none;
  }
`;

function Bookshelf({ books, dispatchRemoveBook }) {
  const history = useHistory();
  const query = new URLSearchParams(useLocation().search);
  const sort = query.get("sort") || "asc";
  const view = query.get("view") || "grid";
  const isSortAsc = sort === "asc";
  const isGridView = view === "grid";

  const [bookModalOpen, setBookModalOpen] = useState(false);

  function openBookModal() {
    setBookModalOpen(true);
  }

  function closeBookModal() {
    setBookModalOpen(false);
  }

  function toggleSort() {
    query.set("sort", isSortAsc ? "desc" : "asc");
    history.push({ pathname: "/", search: `?${query.toString()}` });
  }

  function toggleGrid() {
    query.set("view", isGridView ? "list" : "grid");
    history.push({ pathname: "/", search: `?${query.toString()}` });
  }

  function removeBook(id) {
    dispatchRemoveBook(id);
  }

  function renderBooks() {
    const sortedBooks = books.sort((a, b) => {
      if (isSortAsc) {
        return a.name.toLowerCase() > b.name.toLowerCase();
      } else {
        return a.name.toLowerCase() < b.name.toLowerCase();
      }
    });

    return sortedBooks.map((book) => (
      <StyledLink key={book.id} to={`/book/${book.id}`}>
        <Book
          id={book.id}
          name={book.name}
          author={book.author}
          remove={removeBook}
        />
      </StyledLink>
    ));
  }

  return (
    <div>
      {bookModalOpen && <AddBookModal close={closeBookModal} />}
      <Header>Bookshelf</Header>
      <GridControls>
        <Button variant="text" onClick={toggleSort}>
          {isSortAsc ? <FaChevronDown /> : <FaChevronUp />}
          Sort
        </Button>
        <Button onClick={openBookModal}>
          + <FaBook title="add-book" />
        </Button>
        <Button variant="text" onClick={toggleGrid}>
          {isGridView ? (
            <span>
              List/<strong>Grid</strong>
            </span>
          ) : (
            <span>
              <strong>List</strong>/Grid
            </span>
          )}
        </Button>
      </GridControls>
      <BookList gridView={isGridView}>{renderBooks()}</BookList>
    </div>
  );
}

const mapStateToProps = (state) => ({
  books: state,
});

const mapDispatchToProps = (dispatch) => ({
  dispatchRemoveBook: (id) => dispatch(removeBook(id)),
});

export default connect(mapStateToProps, mapDispatchToProps)(Bookshelf);
