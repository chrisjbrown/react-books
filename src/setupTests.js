import "@testing-library/jest-dom";
import { BrowserRouter } from "react-router-dom";
import { render } from "@testing-library/react";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import "jest-styled-components";

export function renderWithRouter(
  ui,
  { route = "/", initialState = [], ...renderOptions } = {}
) {
  window.history.pushState({}, "Test page", route);
  const mockStore = configureStore([]);
  const store = mockStore(initialState);
  function Wrapper({ children }) {
    return (
      <BrowserRouter>
        <Provider store={store}>{children}</Provider>
      </BrowserRouter>
    );
  }

  return { ...render(ui, { wrapper: Wrapper, ...renderOptions }), store };
}
