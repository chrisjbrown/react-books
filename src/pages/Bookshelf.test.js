import React from "react";
import { renderWithRouter } from "../setupTests";
import { fireEvent } from "@testing-library/react";

import BookShelf from "./BookShelf";

const stateWithBooks = [
  { id: 1, name: "huck finn", author: "mark twain" },
  { id: 2, name: "Lion, witch, and wardrobe", author: "CS Lewis" },
];

describe("BookShelf page", () => {
  it("should have default view", () => {
    const { container } = renderWithRouter(<BookShelf />);

    expect(container.firstChild).toMatchSnapshot();
  });

  it("should have default view with books", () => {
    const { container } = renderWithRouter(<BookShelf />, {
      initialState: stateWithBooks,
    });

    expect(container.firstChild).toMatchSnapshot();
  });

  it("should sort books desc", () => {
    const { container, getByText } = renderWithRouter(<BookShelf />, {
      initialState: stateWithBooks,
    });

    fireEvent.click(getByText("Sort"));

    expect(container.firstChild).toMatchSnapshot();
  });

  it("should render books as list", () => {
    const { container, getByText } = renderWithRouter(<BookShelf />, {
      initialState: stateWithBooks,
    });

    fireEvent.click(getByText("List/"));

    expect(container.firstChild).toMatchSnapshot();
  });

  it("should call add book", () => {
    const { getByLabelText, getByTitle, getByText, store } = renderWithRouter(
      <BookShelf />,
      {
        initialState: stateWithBooks,
      }
    );

    fireEvent.click(getByTitle("add-book"));

    fireEvent.change(getByLabelText("Name:"), {
      target: { value: "my book" },
    });
    fireEvent.change(getByLabelText("Author:"), {
      target: { value: "me" },
    });
    fireEvent.click(getByText("Done"));

    const actions = store.getActions();
    const expectedPayload = {
      type: "ADD_BOOK",
      book: {
        name: "my book",
        author: "me",
        customFields: [],
      },
    };
    expect(actions).toEqual([expectedPayload]);
  });

  it("should call remove book", () => {
    const { getAllByText, store } = renderWithRouter(<BookShelf />, {
      initialState: stateWithBooks,
    });

    fireEvent.click(getAllByText("X")[0]);

    const actions = store.getActions();
    const expectedPayload = { type: "REMOVE_BOOK", id: 1 };
    expect(actions).toEqual([expectedPayload]);
  });
});
