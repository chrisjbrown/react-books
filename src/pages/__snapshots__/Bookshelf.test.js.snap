// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookShelf page should have default view 1`] = `
.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: transparent;
  border: 0;
  padding: 10px 20px;
  cursor: pointer;
  outline: 0;
}

.c2:active {
  background-color: lightblue;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: transparent;
  border: 1px solid lightblue;
  padding: 10px 20px;
  cursor: pointer;
  outline: 0;
}

.c3:active {
  background-color: lightblue;
}

.c0 {
  text-align: center;
}

.c4 {
  display: grid;
  grid-template-columns: repeat(auto-fit,160px);
  -webkit-column-gap: 20px;
  column-gap: 20px;
  row-gap: 20px;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  padding: 10px;
  border: 1px solid black;
  margin-bottom: 20px;
}

<div>
  <h1
    class="c0"
  >
    Bookshelf
  </h1>
  <div
    class="c1"
  >
    <button
      class="c2"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 448 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z"
        />
      </svg>
      Sort
    </button>
    <button
      class="c3"
    >
      + 
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 448 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <title>
          add-book
        </title>
        <path
          d="M448 360V24c0-13.3-10.7-24-24-24H96C43 0 0 43 0 96v320c0 53 43 96 96 96h328c13.3 0 24-10.7 24-24v-16c0-7.5-3.5-14.3-8.9-18.7-4.2-15.4-4.2-59.3 0-74.7 5.4-4.3 8.9-11.1 8.9-18.6zM128 134c0-3.3 2.7-6 6-6h212c3.3 0 6 2.7 6 6v20c0 3.3-2.7 6-6 6H134c-3.3 0-6-2.7-6-6v-20zm0 64c0-3.3 2.7-6 6-6h212c3.3 0 6 2.7 6 6v20c0 3.3-2.7 6-6 6H134c-3.3 0-6-2.7-6-6v-20zm253.4 250H96c-17.7 0-32-14.3-32-32 0-17.6 14.4-32 32-32h285.4c-1.9 17.1-1.9 46.9 0 64z"
        />
      </svg>
    </button>
    <button
      class="c2"
    >
      <span>
        List/
        <strong>
          Grid
        </strong>
      </span>
    </button>
  </div>
  <div
    class="c4"
  />
</div>
`;

exports[`BookShelf page should have default view with books 1`] = `
.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  position: relative;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 20px;
  border: 1px solid pink;
  text-overflow: ellipsis;
  overflow: hidden;
}

.c6 .remove {
  position: absolute;
  top: 0;
  right: 0;
  color: black;
  font-size: 16px;
  padding: 10px;
  border: 1px solid black;
}

.c6 .remove:hover,
.c6 .remove:active,
.c6 .remove:focus {
  background-color: black;
  color: white;
}

.c6 .text {
  margin-top: 20px;
  color: black;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c6 .text > div:last-child {
  margin-top: 10px;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: transparent;
  border: 0;
  padding: 10px 20px;
  cursor: pointer;
  outline: 0;
}

.c2:active {
  background-color: lightblue;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: transparent;
  border: 1px solid lightblue;
  padding: 10px 20px;
  cursor: pointer;
  outline: 0;
}

.c3:active {
  background-color: lightblue;
}

.c0 {
  text-align: center;
}

.c4 {
  display: grid;
  grid-template-columns: repeat(auto-fit,160px);
  -webkit-column-gap: 20px;
  column-gap: 20px;
  row-gap: 20px;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  padding: 10px;
  border: 1px solid black;
  margin-bottom: 20px;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c5:focus,
.c5:hover,
.c5:visited,
.c5:link,
.c5:active {
  -webkit-text-decoration: none;
  text-decoration: none;
}

<div>
  <h1
    class="c0"
  >
    Bookshelf
  </h1>
  <div
    class="c1"
  >
    <button
      class="c2"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 448 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z"
        />
      </svg>
      Sort
    </button>
    <button
      class="c3"
    >
      + 
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 448 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <title>
          add-book
        </title>
        <path
          d="M448 360V24c0-13.3-10.7-24-24-24H96C43 0 0 43 0 96v320c0 53 43 96 96 96h328c13.3 0 24-10.7 24-24v-16c0-7.5-3.5-14.3-8.9-18.7-4.2-15.4-4.2-59.3 0-74.7 5.4-4.3 8.9-11.1 8.9-18.6zM128 134c0-3.3 2.7-6 6-6h212c3.3 0 6 2.7 6 6v20c0 3.3-2.7 6-6 6H134c-3.3 0-6-2.7-6-6v-20zm0 64c0-3.3 2.7-6 6-6h212c3.3 0 6 2.7 6 6v20c0 3.3-2.7 6-6 6H134c-3.3 0-6-2.7-6-6v-20zm253.4 250H96c-17.7 0-32-14.3-32-32 0-17.6 14.4-32 32-32h285.4c-1.9 17.1-1.9 46.9 0 64z"
        />
      </svg>
    </button>
    <button
      class="c2"
    >
      <span>
        List/
        <strong>
          Grid
        </strong>
      </span>
    </button>
  </div>
  <div
    class="c4"
  >
    <a
      class="c5"
      href="/book/1"
    >
      <div
        class="c6"
      >
        <span
          class="remove"
        >
          X
        </span>
        <div
          class="text"
        >
          <div>
            Name: 
            huck finn
          </div>
          <div>
            Author: 
            mark twain
          </div>
        </div>
      </div>
    </a>
    <a
      class="c5"
      href="/book/2"
    >
      <div
        class="c6"
      >
        <span
          class="remove"
        >
          X
        </span>
        <div
          class="text"
        >
          <div>
            Name: 
            Lion, witch, and wardrobe
          </div>
          <div>
            Author: 
            CS Lewis
          </div>
        </div>
      </div>
    </a>
  </div>
</div>
`;
